certora/specs/SocialRecoveryModule.spec

```diff
@@ -1,26 +1,131 @@
 function requireSocialRecoveryModuleEnabled() {
     require(safeContract.isModuleEnabled(currentContract));
 }
 
-// This is a dummy rule to verify the Safe Contract Setup with the Social Recovery
-// Module is working as intended.
-rule recoveryModuleCanBeDisabled {
-    env e;
-    address prevModule;
+// This integrity rule verifies that the addGuardianWithThreshold(...) works as expected.
+rule addGuardianWorksAsExpected(env e, address guardian, uint256 threshold, address otherGuardian) {
+
+    // This can be removed once linked list invariant is implemented.
+    require guardianStorageContract.entries[safeContract].count > 0 => guardianStorageContract.entries[safeContract].guardians[SENTINEL()] != 0; // Sentinel should not be zero if there are guardians.
+
+    require guardian != otherGuardian;
+
+    bool otherWasGuardian = currentContract.isGuardian(safeContract, otherGuardian);
+
+    currentContract.addGuardianWithThreshold(e, safeContract, guardian, threshold);
+
+    assert safeContract.isModuleEnabled(currentContract);
+    assert e.msg.sender == safeContract;
+    assert currentContract.isGuardian(safeContract, guardian);
+    assert otherWasGuardian == currentContract.isGuardian(safeContract, otherGuardian);
+}
+
+// This integrity rule verifies that the guardian can always be added considering ideal conditions.
+rule guardianCanAlwaysBeAdded(env e, address guardian, uint256 threshold, address otherGuardian) {
 
     requireSocialRecoveryModuleEnabled();
 
-    safeContract.disableModule@withrevert(e, prevModule, currentContract);
+    require e.msg.value == 0;
+    require threshold > 0;
+    require guardianStorageContract.entries[safeContract].count < max_uint256; // To prevent overflow (Realistically can't reach).
+
+    // This can be removed once linked list invariant is implemented.
+    require guardianStorageContract.entries[safeContract].count > 0 => guardianStorageContract.entries[safeContract].guardians[SENTINEL()] != 0; // Sentinel should not be zero if there are guardians.
+
+    require guardian != 0;
+    require guardian != SENTINEL();
+    require guardian != safeContract;
+    require guardian != otherGuardian;
+
+    require !safeContract.isOwner(guardian);
+    require !currentContract.isGuardian(safeContract, guardian);
+
+    require threshold < guardianStorageContract.countGuardians(safeContract);
+
+    bool otherWasGuardian = currentContract.isGuardian(safeContract, otherGuardian);
+
+    require e.msg.sender == safeContract;
+    currentContract.addGuardianWithThreshold@withrevert(e, safeContract, guardian, threshold);
+    bool isReverted = lastReverted;
+
+    assert !isReverted => currentContract.isGuardian(safeContract, guardian);
+    assert otherWasGuardian == currentContract.isGuardian(safeContract, otherGuardian);
+}
+
+// This integrity rule verifies that the revokeGuardianWithThreshold(...) works as expected.
```

Comment:
Again, sorry to pick on the comments, but I feel like this doesn't add much value: the function name already says "worksAsExpected".

Can we please clarify what this means in the context of the rule? As in "what is as expected"?

Reply:
Added more details, not sure if that is what you were expecting though.